fuse_optimizers/src/windowed_optimizer.cpp

```diff
@@ -169,6 +175,25 @@ void WindowedOptimizer::optimizationLoop()
       {
         continue;
       }
+      // check if new transaction has added constraints to variables that are to be marginalized
+      std::vector<fuse_core::UUID> faulty_constraints;
+      for(auto& c: new_transaction->addedConstraints()){
+        for(auto var_uuid: c.variables()){
+          for (auto marginal_uuid : marginal_transaction_.removedVariables()) {
+            if (var_uuid == marginal_uuid) {
+              faulty_constraints.push_back(c.uuid());
+              break;
+            }
+          }
+        }
+      }
+      if(faulty_constraints.size() > 0){
+        ROS_WARN_STREAM("Removing invalid constraints.");
+        for(auto& faulty_constraint: faulty_constraints){
+          new_transaction->removeConstraint(faulty_constraint);
+        }
+      }
```

Comment:
Interesting. There is indeed a timing problem. At the end of Loop1 we compute a set of variables to marginalize. But those variables are not removed until the graph is updated in Loop2. The sensor models have no way of knowing a variable was removed in Loop1 until after Loop2 has been processed and the updated graph is sent back to the sensor models. So a sensor model can very reasonably send new constraints involving variables that should have been removed.

I originally implemented it this way to avoid optimizing the graph twice. Once when new constraints are added, and a second time to remove the marginalized variables. In retrospect, marginalizing variables from the graph should not alter the optimized values of the remaining variables. So I could just remove the marginalized variables at the end of the loop without reoptimizing. But I wouldn't want to make that change without a lot of testing first.

For now, your fix should be valid. Thanks.